README
      This is a ISC DHCP Distribution version 4.1.0 customized by Vyatta.
	      Internet Systems Consortium DHCP Distribution
		             Version 4.1.0p1
			      22 June 2009

			      README FILE

You should read this file carefully before trying to install or use
the ISC DHCP Distribution.

			  TABLE OF CONTENTS

	1	WHERE TO FIND DOCUMENTATION
	2	RELEASE STATUS
	3	BUILDING THE DHCP DISTRIBUTION
	 3.1	 UNPACKING IT
	 3.2	 CONFIGURING IT
	  3.2.1	  DYNAMIC DNS UPDATES
	  3.2.2   LOCALLY DEFINED OPTIONS
	 3.3	 BUILDING IT
	4	INSTALLING THE DHCP DISTRIBUTION
	5	USING THE DHCP DISTRIBUTION
	 5.1	  FIREWALL RULES
	 5.2	 LINUX
	  5.2.1	  IF_TR.H NOT FOUND
	  5.2.2	  SO_ATTACH_FILTER UNDECLARED
	  5.2.3	  PROTOCOL NOT CONFIGURED
	  5.2.4	  BROADCAST
	  5.2.6	  IP BOOTP AGENT
	  5.2.7	  MULTIPLE INTERFACES
	 5.3	 SCO
	 5.4	 HP-UX
	 5.5	 ULTRIX
	 5.6	 FreeBSD
	 5.7	 NeXTSTEP
	 5.8	 SOLARIS
	 5.9	 AIX
	 5.10	 MacOS X
	6	SUPPORT
	 6.1	 HOW TO REPORT BUGS

		      WHERE TO FIND DOCUMENTATION

Documentation for this software includes this README file, the
RELNOTES file, and the manual pages, which are in the server, common,
client and relay subdirectories.  The README file (this file) includes
late-breaking operational and system-specific information that you
should read even if you don't want to read the manual pages, and that
you should *certainly* read if you run into trouble.  Internet
standards relating to the DHCP protocol are stored in the doc
subdirectory.  You will have the best luck reading the manual pages if
you build this software and then install it, although you can read
them directly out of the distribution if you need to.

DHCP server documentation is in the dhcpd man page.  Information about
the DHCP server lease database is in the dhcpd.leases man page.
Server configuration documentation is in the dhcpd.conf man page as
well as the dhcp-options man page.   A sample DHCP server
configuration is in the file server/dhcpd.conf.   The source for the
dhcpd, dhcpd.leases and dhcpd.conf man pages is in the server/ sub-
directory in the distribution.   The source for the dhcp-options.5
man page is in the common/ subdirectory.

DHCP Client documentation is in the dhclient man page.  DHCP client
configuration documentation is in the dhclient.conf man page and the
dhcp-options man page.  The DHCP client configuration script is
documented in the dhclient-script man page.   The format of the DHCP
client lease database is documented in the dhclient.leases man page.
The source for all these man pages is in the client/ subdirectory in
the distribution.   In addition, the dhcp-options man page should be
referred to for information about DHCP options.

DHCP relay agent documentation is in the dhcrelay man page, the source
for which is distributed in the relay/ subdirectory.

To read installed manual pages, use the man command.  Type "man page"
where page is the name of the manual page.   This will only work if
you have installed the ISC DHCP distribution using the ``make install''
command (described later).

If you want to read manual pages that aren't installed, you can type
``nroff -man page |more'' where page is the filename of the
unformatted manual page.  The filename of an unformatted manual page
is the name of the manual page, followed by '.', followed by some
number - 5 for documentation about files, and 8 for documentation
about programs.   For example, to read the dhcp-options man page,
you would type ``nroff -man common/dhcp-options.5 |more'', assuming
your current working directory is the top level directory of the ISC
DHCP Distribution.

Please note that the pathnames of files to which our manpages refer
will not be correct for your operating system until after you iterate
'make install' (so if you're reading a manpage out of the source
directory, it may not have up-to-date information).

			    RELEASE STATUS

This is ISC DHCP 4.1.0p1, a PATCH release which addresses a security
problem with the DHCPv4 client ('dhclient') included in ISC DHCP 4.1.0.

In this release, the DHCPv6 server should be fully functional on Linux,
Solaris, or any BSD.  The DHCPv6 client should be similarly functional
except on Solaris.

The DHCPv4 server, relay, and client, should be fully functional
on Linux, Solaris, any BSD, HPUX, SCO, NextSTEP, and Irix.

If you are running the DHCP distribution on a machine which is a
firewall, or if there is a firewall between your DHCP server(s) and
DHCP clients, please read the section on firewalls which appears later
in this document.

If you wish to run the DHCP Distribution on Linux, please see the
Linux-specific notes later in this document.  If you wish to run on an
SCO release, please see the SCO-specific notes later in this document.
You particularly need to read these notes if you intend to support
Windows 95 clients.  If you are running a version of FreeBSD prior to
2.2, please read the note on FreeBSD.  If you are running HP-UX or
Ultrix, please read the notes for those operating systems below.  If
you are running NeXTSTEP, please see the notes on NeXTSTEP below.

If you start dhcpd and get a message, "no free bpf", that means you
need to configure the Berkeley Packet Filter into your operating
system kernel.   On NetBSD, FreeBSD and BSD/os, type ``man bpf'' for
information.   On Digital Unix, type ``man pfilt''.


		    BUILDING THE DHCP DISTRIBUTION

			     UNPACKING IT

To build the DHCP Distribution, unpack the compressed tar file using
the tar utility and the gzip command - type something like:

	gunzip dhcp-4.1.0p1.tar.gz
	tar xvf dhcp-4.1.0p1.tar

			    CONFIGURING IT

Now, cd to the dhcp-4.1.0p1 subdirectory that you've just created and
configure the source tree by typing:

	./configure

If the configure utility can figure out what sort of system you're
running on, it will create a custom Makefile for you for that
system; otherwise, it will complain.  If it can't figure out what
system you are using, that system is not supported - you are on
your own.

			 DYNAMIC DNS UPDATES

A fully-featured implementation of dynamic DNS updates is included in
this release.   There are no build dependencies with any BIND version
- this version can and should just use the resolver in your C library.

There is documentation for the DDNS support in the dhcpd.conf manual
page - see the beginning of this document for information on finding
manual pages.

		       LOCALLY DEFINED OPTIONS

In previous versions of the DHCP server there was a mechanism whereby
options that were not known by the server could be configured using
a name made up of the option code number and an identifier:
"option-nnn"   This is no longer supported, because it is not future-
proof.   Instead, if you want to use an option that the server doesn't
know about, you must explicitly define it using the method described
in the dhcp-options man page under the DEFINING NEW OPTIONS heading.

			     BUILDING IT

Once you've run configure, just type ``make'', and after a while
you should have a dhcp server.  If you get compile errors on one
of the supported systems mentioned earlier, please let us know.
If you get warnings, it's not likely to be a problem - the DHCP
server compiles completely warning-free on as many architectures
as we can manage, but there are a few for which this is difficult.
If you get errors on a system not mentioned above, you will need
to do some programming or debugging on your own to get the DHCP
Distribution working.

		   INSTALLING THE DHCP DISTRIBUTION

Once you have successfully gotten the DHCP Distribution to build, you
can install it by typing ``make install''.   If you already have an old
version of the DHCP Distribution installed, you may want to save it
before typing ``make install''.

		     USING THE DHCP DISTRIBUTION

			    FIREWALL RULES

If you are running the DHCP server or client on a computer that's also
acting as a firewall, you must be sure to allow DHCP packets through
the firewall.  In particular, your firewall rules _must_ allow packets
from IP address 0.0.0.0 to IP address 255.255.255.255 from UDP port 68
to UDP port 67 through.  They must also allow packets from your local
firewall's IP address and UDP port 67 through to any address your DHCP
server might serve on UDP port 68.  Finally, packets from relay agents
on port 67 to the DHCP server on port 67, and vice versa, must be
permitted.

We have noticed that on some systems where we are using a packet
filter, if you set up a firewall that blocks UDP port 67 and 68
entirely, packets sent through the packet filter will not be blocked.
However, unicast packets will be blocked.   This can result in strange
behaviour, particularly on DHCP clients, where the initial packet
exchange is broadcast, but renewals are unicast - the client will
appear to be unable to renew until it starts broadcasting its
renewals, and then suddenly it'll work.   The fix is to fix the
firewall rules as described above.

			   PARTIAL SERVERS

If you have a server that is connected to two networks, and you only
want to provide DHCP service on one of those networks (e.g., you are
using a cable modem and have set up a NAT router), if you don't write
any subnet declaration for the network you aren't supporting, the DHCP
server will ignore input on that network interface if it can.  If it
can't, it will refuse to run - some operating systems do not have the
capability of supporting DHCP on machines with more than one
interface, and ironically this is the case even if you don't want to
provide DHCP service on one of those interfaces.

				LINUX

There are three big LINUX issues: the all-ones broadcast address,
Linux 2.1 ip_bootp_agent enabling, and operations with more than one
network interface.   There are also two potential compilation/runtime
problems for Linux 2.1/2.2: the "SO_ATTACH_FILTER undeclared" problem
and the "protocol not configured" problem.

		    LINUX: PROTOCOL NOT CONFIGURED

If you get the following message, it's because your kernel doesn't
have the linux packetfilter or raw packet socket configured:

 Make sure CONFIG_PACKET (Packet socket) and CONFIG_FILTER (Socket
 Filtering) are enabled in your kernel configuration

If this happens, you need to configure your Linux kernel to support
Socket Filtering and the Packet socket, or to select a kernel provided
by your Linux distribution that has these enabled (virtually all modern
ones do by default).

			   LINUX: BROADCAST

If you are running a recent version of Linux, this won't be a problem,
but on older versions of Linux (kernel versions prior to 2.2), there
is a potential problem with the broadcast address being sent
incorrectly.

In order for dhcpd to work correctly with picky DHCP clients (e.g.,
Windows 95), it must be able to send packets with an IP destination
address of 255.255.255.255.  Unfortunately, Linux changes an IP
destination of 255.255.255.255 into the local subnet broadcast address
(here, that's 192.5.5.223).

This isn't generally a problem on Linux 2.2 and later kernels, since
we completely bypass the Linux IP stack, but on old versions of Linux
2.1 and all versions of Linux prior to 2.1, it is a problem - pickier
DHCP clients connected to the same network as the ISC DHCP server or
ISC relay agent will not see messages from the DHCP server.   It *is*
possible to run into trouble with this on Linux 2.2 and later if you
are running a verson of the DHCP server that was compiled on a Linux
2.0 system, though.

It is possible to work around this problem on some versions of Linux
by creating a host route from your network interface address to
255.255.255.255.   The command you need to use to do this on Linux
varies from version to version.   The easiest version is:

	route add -host 255.255.255.255 dev eth0

On some older Linux systems, you will get an error if you try to do
this.   On those systems, try adding the following entry to your
/etc/hosts file:

255.255.255.255	all-ones

Then, try:

	route add -host all-ones dev eth0

Another route that has worked for some users is:

	route add -net 255.255.255.0 dev eth0

If you are not using eth0 as your network interface, you should
specify the network interface you *are* using in your route command.

			LINUX: IP BOOTP AGENT

Some versions of the Linux 2.1 kernel apparently prevent dhcpd from
working unless you enable it by doing the following:

	      echo 1 >/proc/sys/net/ipv4/ip_bootp_agent


		      LINUX: MULTIPLE INTERFACES

Very old versions of the Linux kernel do not provide a networking API
that allows dhcpd to operate correctly if the system has more than one
broadcast network interface.  However, Linux 2.0 kernels with version
numbers greater than or equal to 2.0.31 add an API feature: the
SO_BINDTODEVICE socket option.  If SO_BINDTODEVICE is present, it is
possible for dhcpd to operate on Linux with more than one network
interface.  In order to take advantage of this, you must be running a
2.0.31 or greater kernel, and you must have 2.0.31 or later system
headers installed *before* you build the DHCP Distribution.

We have heard reports that you must still add routes to 255.255.255.255
in order for the all-ones broadcast to work, even on 2.0.31 kernels.
In fact, you now need to add a route for each interface.   Hopefully
the Linux kernel gurus will get this straight eventually.

Linux 2.1 and later kernels do not use SO_BINDTODEVICE or require the
broadcast address hack, but do support multiple interfaces, using the
Linux Packet Filter.

			     LINUX: OpenWrt

DHCP 4.1 has been tested on OpenWrt 7.09 and 8.09.  In keeping with
standard practice, client/scripts now includes a dhclient-script file
for OpenWrt.  However, this is not sufficient by itself to run dhcp on
OpenWrt; a full OpenWrt package for DHCP is available at
ftp://ftp.isc.org/isc/dhcp/dhcp-4.1.0-openwrt.tar.gz

		    LINUX: 802.1q VLAN INTERFACES

If you're using 802.1q vlan interfaces on Linux, it is necessary to
vconfig the subinterface(s) to rewrite the 802.1q information out of
packets received by the dhcpd daemon via LPF:

	vconfig set_flag eth1.523 1 1

Note that this may affect the performance of your system, since the
Linux kernel must rewrite packets received via this interface.  For
more information, consult the vconfig man pages.

				 SCO

ISC DHCP will now work correctly on newer versions of SCO out of the
box (tested on OpenServer 5.05b, assumed to work on UnixWare 7).

Older versions of SCO have the same problem as Linux (described earlier).
The thing is, SCO *really* doesn't want to let you add a host route to
the all-ones broadcast address.

You can try the following:

  ifconfig net0 xxx.xxx.xxx.xxx netmask 0xNNNNNNNN broadcast 255.255.255.255

If this doesn't work, you can also try the following strange hack:

  ifconfig net0 alias 10.1.1.1 netmask 8.0.0.0

Apparently this works because of an interaction between SCO's support
for network classes and the weird netmask.  The 10.* network is just a
dummy that can generally be assumed to be safe.   Don't ask why this
works.   Just try it.   If it works for you, great.

				HP-UX

HP-UX has the same problem with the all-ones broadcast address that
SCO and Linux have.   One user reported that adding the following to
/etc/rc.config.d/netconf helped (you may have to modify this to suit
your local configuration):

INTERFACE_NAME[0]=lan0
IP_ADDRESS[0]=1.1.1.1
SUBNET_MASK[0]=255.255.255.0
BROADCAST_ADDRESS[0]="255.255.255.255"
LANCONFIG_ARGS[0]="ether"
DHCP_ENABLE[0]=0

				ULTRIX

Now that we have Ultrix packet filter support, the DHCP Distribution
on Ultrix should be pretty trouble-free.  However, one thing you do
need to be aware of is that it now requires that the pfilt device be
configured into your kernel and present in /dev.  If you type ``man
packetfilter'', you will get some information on how to configure your
kernel for the packet filter (if it isn't already) and how to make an
entry for it in /dev.

			       FreeBSD

Versions of FreeBSD prior to 2.2 have a bug in BPF support in that the
ethernet driver swaps the ethertype field in the ethernet header
downstream from BPF, which corrupts the output packet.   If you are
running a version of FreeBSD prior to 2.2, and you find that dhcpd
can't communicate with its clients, you should #define BROKEN_FREEBSD_BPF 
in site.h and recompile.

Modern versions of FreeBSD include the ISC DHCP 3.0 client as part of
the base system, and the full distribution (for the DHCP server and
relay agent) is available from the Ports Collection in
/usr/ports/net/isc-dhcp3, or as a package on FreeBSD installation
CDROMs.

                              NeXTSTEP

The NeXTSTEP support uses the NeXTSTEP Berkeley Packet Filter
extension, which is not included in the base NextStep system.  You
must install this extension in order to get dhcpd or dhclient to work.

			       SOLARIS

One problem which has been observed and is not fixed in this
patchlevel has to do with using DLPI on Solaris machines.  The symptom
of this problem is that the DHCP server never receives any requests.

debian/changelog
vyatta-dhcp3 (4.1.8+vyos2+current1) unstable; urgency=medium

  [ Thomas Jepp ]
  * Add Provides to vyatta-dhcp3 to stop it conflicting with isc-dhcp packages.

  [ Kim Hagen ]

 -- Kim Hagen <kim.sidney@gmail.com>  Sun, 24 Jan 2016 15:10:49 -0500

vyatta-dhcp3 (4.1.8+vyos2+lithium5) unstable; urgency=low

  * vyatta-dhcp3: update dh_gencontrol with new development build flag

 -- Alex Harpin <development@landsofshadow.co.uk>  Thu, 18 Jun 2015 20:37:38 +0100
 
vyatta-dhcp3 (4.1.8+vyos2+lithium4) unstable; urgency=low

  [ Dominik Malis ]
  * Bug 520 - Feature request: Show DUID of DHCPv6 client request in log

 -- Alex Harpin <development@landsofshadow.co.uk>  Sat, 13 Jun 2015 10:38:50 +0100
 
vyatta-dhcp3 (4.1.8+vyos2+lithium3) unstable; urgency=low

  [ Alex Harpin ]
  * vyatta-dhcp3: call make_resolv_conf after the ip address is active

 -- Alex Harpin <development@landsofshadow.co.uk>  Sat, 28 Feb 2015 11:44:17 +0000
 
vyatta-dhcp3 (4.1.8+vyos2+lithium2) unstable; urgency=low

  * Update maintainer address

 -- Alex Harpin <development@landsofshadow.co.uk>  Fri, 26 Dec 2014 11:01:18 +0000
 
vyatta-dhcp3 (4.1.8+vyos2+lithium1) unstable; urgency=low

  * New branch

 -- Alex Harpin <development@landsofshadow.co.uk>  Fri, 05 Dec 2014 19:43:48 +0000
 
vyatta-dhcp3 (4.1.8+vyos1+helium2) unstable; urgency=low

  * Replace bash with vbash in client script.

 -- Daniil Baturin <daniil@baturin.org>  Fri, 26 Sep 2014 01:06:11 +0200

vyatta-dhcp3 (4.1.8+vyos1+helium1) unstable; urgency=low

  * New branch

 -- Daniil Baturin <daniil@baturin.org>  Sat, 15 Feb 2014 18:17:38 +0100

vyatta-dhcp3 (4.1.8+hydrogen3) unstable; urgency=low

  [ Mihail Vasiliev ]
  * Update to ISC DHCP 4.1-ESV-R8
  * Vyatta-marked edits re-added.
  * Versioning and maintainers fixup.
  * Last changes by vyatta.
  * Fix printing of long as int.

  [ Daniil Baturin ]

 -- Daniil Baturin <daniil@baturin.org>  Thu, 28 Nov 2013 16:41:54 +0100

vyatta-dhcp3 (4.1.8+hydrogen2) unstable; urgency=low

  * ISC DHCP version 4.1-ESV-R8

 -- Mihail Vasiliev <mickvav@gmail.com>  Wed, 27 Nov 2013 17:08:06 +0400

vyatta-dhcp3 (4.1.2+vyatta11) unstable; urgency=low

  * New branch

 -- Daniil Baturin <daniil@baturin.org>  Sun, 17 Nov 2013 00:25:00 +0100

vyatta-dhcp3 (4.1.2+vyatta9+daisy2) unstable; urgency=low

  * allow optional dependency on vyatta-unicast

 -- Robert Bays <robert@vyatta.com>  Sat, 01 Jun 2013 16:53:06 -0700

vyatta-dhcp3 (4.1.2+vyatta9+daisy1) unstable; urgency=low

  * new branch

 -- James Davidson <james.davidson@vyatta.com>  Mon, 14 Jan 2013 17:07:06 -0800

vyatta-dhcp3 (4.1.2+vyatta9) unstable; urgency=low

  * ISC DHCP version 4.1-ESV-R7
  * Update .gitignore

 -- James Davidson <james.davidson@vyatta.com>  Mon, 17 Sep 2012 11:35:52 -0700

vyatta-dhcp3 (4.1.2+vyatta8) unstable; urgency=low

  * ISC DHCP version 4.1-ESV-R6
  * Ignore .deps directories

 -- James Davidson <james.davidson@vyatta.com>  Mon, 30 Jul 2012 14:29:37 -0700

vyatta-dhcp3 (4.1.2+vyatta7) unstable; urgency=low

  * new branch

 -- Deepti Kulkarni <deepti@vyatta.com>  Sat, 03 Mar 2012 02:25:53 -0800

vyatta-dhcp3 (4.1.2+vyatta6) unstable; urgency=low

  * ISC DHCP version 4.1-ESV-R3

 -- Bob Gilligan <gilligan@vyatta.com>  Thu, 25 Aug 2011 16:33:42 -0700

vyatta-dhcp3 (4.1.2+vyatta5) unstable; urgency=low

  * new branch

 -- Deepti Kulkarni <deepti@vyatta.com>  Thu, 07 Jul 2011 20:56:43 -0700

vyatta-dhcp3 (4.1.2+vyatta4) unstable; urgency=low

  * ISC DHCP version 4.1.1-P1
  * ISC DHCP version 4.1.2
  * ISC DHCP 4.1.2-P1

 -- Bob Gilligan <gilligan@vyatta.com>  Fri, 25 Mar 2011 01:45:54 -0700

vyatta-dhcp3 (4.1.2+vyatta3) unstable; urgency=low

  * Merge branch 'upstream' into mendocino
  * Merged ISC DHCP version 4.1.2.

 -- Bob Gilligan <gilligan@vyatta.com>  Tue, 25 Jan 2011 18:04:55 -0800

vyatta-dhcp3 (4.1.2+vyatta2) unstable; urgency=low

  * Merged ISC DHCP Version 4.1.2.

 -- Bob Gilligan <gilligan@vyatta.com>  Tue, 25 Jan 2011 18:00:23 -0800

vyatta-dhcp3 (4.1.1+vyatta7) unstable; urgency=low

  * don't ignore .*

 -- An-Cheng Huang <ancheng@vyatta.com>  Tue, 28 Dec 2010 13:48:55 -0800

vyatta-dhcp3 (4.1.1+vyatta6) unstable; urgency=low

  * UNRELEASED

 -- An-Cheng Huang <ancheng@vyatta.com>  Thu, 02 Sep 2010 18:28:59 -0700

vyatta-dhcp3 (4.1.1+vyatta5) unstable; urgency=low

  * fix squeeze build and clean up
  * more cleanup

 -- An-Cheng Huang <ancheng@vyatta.com>  Wed, 01 Sep 2010 10:54:56 -0700

vyatta-dhcp3 (4.1.1+vyatta4) unstable; urgency=low

  * ISC DHCP version 4.1.1-P1

 -- Bob Gilligan <gilligan@vyatta.com>  Tue, 27 Jul 2010 20:13:21 -0700

vyatta-dhcp3 (4.1.1+vyatta3) unstable; urgency=low

  * UNRELEASED

 -- An-Cheng Huang <ancheng@vyatta.com>  Thu, 22 Jul 2010 17:24:14 -0700

vyatta-dhcp3 (4.1.1+vyatta2) unstable; urgency=low

  * Add hooks to allow DHCPv6 client parameters only mode.

 -- Bob Gilligan <gilligan@vyatta.com>  Mon, 31 May 2010 23:56:25 -0700

vyatta-dhcp3 (4.1.1+vyatta1) unstable; urgency=low

  * UNRELEASED

 -- Bob Gilligan <gilligan@vyatta.com>  Thu, 20 May 2010 16:42:55 -0700

vyatta-dhcp3 (4.1.0+vyatta16) unstable; urgency=low

  [ Mohit Mehta ]
  * merge 4.1.0p1

  [ Bob Gilligan ]
  * ISC DHCP version 4.1.1.

 -- Bob Gilligan <gilligan@vyatta.com>  Thu, 20 May 2010 16:37:03 -0700

vyatta-dhcp3 (4.1.0+vyatta15) unstable; urgency=low

  * Don't flush all perminent IPv6 addresses from interface when
    renewing.

 -- Bob Gilligan <gilligan@vyatta.com>  Wed, 21 Apr 2010 15:06:10 -0700

vyatta-dhcp3 (4.1.0+vyatta14) unstable; urgency=low

  * Update DHCP client back-end script for DHCPv6.

 -- Bob Gilligan <gilligan@vyatta.com>  Fri, 16 Apr 2010 16:35:34 -0700

vyatta-dhcp3 (4.1.0+vyatta13) unstable; urgency=low

  * Force dependency on recent version of quagga

 -- Stephen Hemminger <stephen.hemminger@vyatta.com>  Mon, 08 Mar 2010 16:53:39 -0800

vyatta-dhcp3 (4.1.0+vyatta12) unstable; urgency=low

  * Use vtysh not vyatta-vtysh


dst/base64.c
/*
 * Copyright (c) 2004,2009 by Internet Systems Consortium, Inc. ("ISC")
 * Copyright (c) 1996-2003 by Internet Software Consortium
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND ISC DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS.  IN NO EVENT SHALL ISC BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
 * OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 *   Internet Systems Consortium, Inc.
 *   950 Charter Street
 *   Redwood City, CA 94063
 *   <info@isc.org>
 *   https://www.isc.org/
 */

/*
 * Portions Copyright (c) 1995 by International Business Machines, Inc.
 *
 * International Business Machines, Inc. (hereinafter called IBM) grants
 * permission under its copyrights to use, copy, modify, and distribute this
 * Software with or without fee, provided that the above copyright notice and
 * all paragraphs of this notice appear in all copies, and that the name of IBM
 * not be used in connection with the marketing of any product incorporating
 * the Software or modifications thereof, without specific, written prior
 * permission.
 *
 * To the extent it has a right to do so, IBM grants an immunity from suit
 * under its patents, if any, for the use, sale or manufacture of products to
 * the extent that such products are used for performing Domain Name System
 * dynamic updates in TCP/IP networks by means of the Software.  No immunity is
 * granted for any product per se or for any other function of any product.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", AND IBM DISCLAIMS ALL WARRANTIES,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE.  IN NO EVENT SHALL IBM BE LIABLE FOR ANY SPECIAL,
 * DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER ARISING
 * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE, EVEN
 * IF IBM IS APPRISED OF THE POSSIBILITY OF SUCH DAMAGES.
 */

#if !defined(LINT) && !defined(CODECENTER)
static const char rcsid[] = "$Id: base64.c,v 1.2.786.2 2009-07-24 22:04:52 sar Exp $";
#endif /* not lint */

#include <sys/types.h>
#include <sys/param.h>
#include <sys/socket.h>

#include <netinet/in.h>
#include <arpa/inet.h>

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <sys/socket.h>

#include "minires/minires.h"
#include "arpa/nameser.h"

#define Assert(Cond) if (!(Cond)) abort()

static const char Base64[] =
	"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
static const char Pad64 = '=';

/* (From RFC1521 and draft-ietf-dnssec-secext-03.txt)
   The following encoding technique is taken from RFC 1521 by Borenstein
   and Freed.  It is reproduced here in a slightly edited form for
   convenience.

   A 65-character subset of US-ASCII is used, enabling 6 bits to be
   represented per printable character. (The extra 65th character, "=",
   is used to signify a special processing function.)

   The encoding process represents 24-bit groups of input bits as output
   strings of 4 encoded characters. Proceeding from left to right, a
   24-bit input group is formed by concatenating 3 8-bit input groups.
   These 24 bits are then treated as 4 concatenated 6-bit groups, each
   of which is translated into a single digit in the base64 alphabet.

   Each 6-bit group is used as an index into an array of 64 printable
   characters. The character referenced by the index is placed in the
   output string.

                         Table 1: The Base64 Alphabet

      Value Encoding  Value Encoding  Value Encoding  Value Encoding
          0 A            17 R            34 i            51 z
          1 B            18 S            35 j            52 0
          2 C            19 T            36 k            53 1
          3 D            20 U            37 l            54 2
          4 E            21 V            38 m            55 3
          5 F            22 W            39 n            56 4
          6 G            23 X            40 o            57 5
          7 H            24 Y            41 p            58 6
          8 I            25 Z            42 q            59 7
          9 J            26 a            43 r            60 8
         10 K            27 b            44 s            61 9
         11 L            28 c            45 t            62 +
         12 M            29 d            46 u            63 /
         13 N            30 e            47 v
         14 O            31 f            48 w         (pad) =
         15 P            32 g            49 x
         16 Q            33 h            50 y

   Special processing is performed if fewer than 24 bits are available
   at the end of the data being encoded.  A full encoding quantum is
   always completed at the end of a quantity.  When fewer than 24 input
   bits are available in an input group, zero bits are added (on the
   right) to form an integral number of 6-bit groups.  Padding at the
   end of the data is performed using the '=' character.

   Since all base64 input is an integral number of octets, only the
         -------------------------------------------------                       
   following cases can arise:
   
       (1) the final quantum of encoding input is an integral
           multiple of 24 bits; here, the final unit of encoded
	   output will be an integral multiple of 4 characters
	   with no "=" padding,
       (2) the final quantum of encoding input is exactly 8 bits;
           here, the final unit of encoded output will be two
	   characters followed by two "=" padding characters, or
       (3) the final quantum of encoding input is exactly 16 bits;
           here, the final unit of encoded output will be three
	   characters followed by one "=" padding character.
   */

int
b64_ntop(u_char const *src, size_t srclength, char *target, size_t targsize) {
	size_t datalength = 0;
	u_char input[3];
	u_char output[4];
	size_t i;

	while (2 < srclength) {
		input[0] = *src++;
		input[1] = *src++;
		input[2] = *src++;
		srclength -= 3;

		output[0] = input[0] >> 2;
		output[1] = ((input[0] & 0x03) << 4) + (input[1] >> 4);
		output[2] = ((input[1] & 0x0f) << 2) + (input[2] >> 6);
		output[3] = input[2] & 0x3f;
		Assert(output[0] < 64);
		Assert(output[1] < 64);
		Assert(output[2] < 64);
		Assert(output[3] < 64);

		if (datalength + 4 > targsize)
			return (-1);
		target[datalength++] = Base64[output[0]];
		target[datalength++] = Base64[output[1]];
		target[datalength++] = Base64[output[2]];
		target[datalength++] = Base64[output[3]];
	}
    
	/* Now we worry about padding. */
	if (0 != srclength) {
		/* Get what's left. */
		input[0] = input[1] = input[2] = '\0';
		for (i = 0; i < srclength; i++)
			input[i] = *src++;
	
		output[0] = input[0] >> 2;
		output[1] = ((input[0] & 0x03) << 4) + (input[1] >> 4);
		output[2] = ((input[1] & 0x0f) << 2) + (input[2] >> 6);
		Assert(output[0] < 64);
		Assert(output[1] < 64);
		Assert(output[2] < 64);

		if (datalength + 4 > targsize)
			return (-1);
		target[datalength++] = Base64[output[0]];
		target[datalength++] = Base64[output[1]];
		if (srclength == 1)
			target[datalength++] = Pad64;
		else
			target[datalength++] = Base64[output[2]];
		target[datalength++] = Pad64;
	}
	if (datalength >= targsize)
		return (-1);
	target[datalength] = '\0';	/* Returned value doesn't count \0. */
	return (datalength);
}

/* skips all whitespace anywhere.
   converts characters, four at a time, starting at (or after)
   src from base - 64 numbers into three 8 bit bytes in the target area.
   it returns the number of data bytes stored at the target, or -1 on error.
 */

int
b64_pton(src, target, targsize)
	char const *src;
	u_char *target;
	size_t targsize;
{
	int tarindex, state, ch;
	char *pos;

	state = 0;
	tarindex = 0;

	while ((ch = *src++) != '\0') {
		if (isspace(ch))	/* Skip whitespace anywhere. */
			continue;

		if (ch == Pad64)
			break;

		pos = strchr(Base64, ch);
		if (pos == 0) 		/* A non-base64 character. */
			return (-1);

		switch (state) {
		case 0:
			if (target) {
				if ((size_t)tarindex >= targsize)
					return (-1);
				target[tarindex] = (pos - Base64) << 2;
			}
			state = 1;
			break;
		case 1:
			if (target) {
				if ((size_t)tarindex + 1 >= targsize)
					return (-1);
				target[tarindex]   |=  (pos - Base64) >> 4;
				target[tarindex+1]  = ((pos - Base64) & 0x0f)
							<< 4 ;
			}
			tarindex++;
			state = 2;
			break;
		case 2:
			if (target) {
				if ((size_t)tarindex + 1 >= targsize)
					return (-1);
				target[tarindex]   |=  (pos - Base64) >> 2;
				target[tarindex+1]  = ((pos - Base64) & 0x03)
							<< 6;
			}
			tarindex++;
			state = 3;
			break;
		case 3:
			if (target) {
				if ((size_t)tarindex >= targsize)
					return (-1);
				target[tarindex] |= (pos - Base64);
			}
			tarindex++;
			state = 0;
			break;
		default:
			abort();
		}
	}

	/*
	 * We are done decoding Base-64 chars.  Let's see if we ended
	 * on a byte boundary, and/or with erroneous trailing characters.
	 */

	if (ch == Pad64) {		/* We got a pad char. */
		ch = *src++;		/* Skip it, get next. */
		switch (state) {
		case 0:		/* Invalid = in first position */
		case 1:		/* Invalid = in second position */
			return (-1);

		case 2:		/* Valid, means one byte of info */
			/* Skip any number of spaces. */
			for ((void)NULL; ch != '\0'; ch = *src++)
				if (!isspace(ch))
					break;
			/* Make sure there is another trailing = sign. */
			if (ch != Pad64)
				return (-1);
			ch = *src++;		/* Skip the = */
			/* Fall through to "single trailing =" case. */
			/* FALLTHROUGH */

		case 3:		/* Valid, means two bytes of info */
			/*
			 * We know this char is an =.  Is there anything but
			 * whitespace after it?
			 */
			for ((void)NULL; ch != '\0'; ch = *src++)
				if (!isspace(ch))
					return (-1);

			/*
			 * Now make sure for cases 2 and 3 that the "extra"
			 * bits that slopped past the last full byte were
			 * zeros.  If we don't check them, they become a
			 * subliminal channel.
			 */
			if (target && target[tarindex] != 0)
				return (-1);
		}
	} else {
		/*
		 * We ended by seeing the end of the string.  Make sure we
		 * have no partial bytes lying around.
		 */
		if (state != 0)
			return (-1);
	}

	return (tarindex);
}
